hare/broker.go

```diff
@@ -327,124 +267,61 @@ func (b *Broker) HandleEligibility(ctx context.Context, em *types.HareEligibilit
 	return false
 }
 
-func (b *Broker) getInbox(id types.LayerID) chan any {
-	b.mu.RLock()
-	defer b.mu.RUnlock()
-	out, exist := b.outbox[id.Uint32()]
-	if !exist {
-		return nil
-	}
-	return out
-}
-
-func (b *Broker) handleEarlyMessage(logger log.Log, layer types.LayerID, nodeID types.NodeID, msg any) error {
+// Register a layer to receive messages
+// Note: the registering instance is assumed to be started and accepting messages.
+func (b *Broker) Register(ctx context.Context, id types.LayerID) (chan any,
+	error,
+) {
 	b.mu.Lock()
 	defer b.mu.Unlock()
-	layerNum := layer.Uint32()
-	if _, exist := b.pending[layerNum]; !exist { // create buffer if first msg
-		b.pending[layerNum] = make([]any, 0, inboxCapacity)
+	if !id.After(b.latestLayer) { // should expect to update only newer layers
+		b.WithContext(ctx).With().Error("tried to update a previous layer",
+			log.Stringer("this_layer", id),
+			log.Stringer("prev_layer", b.latestLayer))
 	}
 
-	// we want to write all buffered messages to a chan with InboxCapacity len
-	// hence, we limit the buffer for pending messages
-	if len(b.pending[layerNum]) == inboxCapacity {
-		logger.With().Warning("too many pending messages, ignoring message",
-			log.Int("inbox_capacity", inboxCapacity),
-			log.Stringer("smesher", nodeID))
-		return nil
+	// We check here since early messages could have already constructed this map entry
+	outboxCh, exist := b.outbox[id.Uint32()]
+	if !exist {
+		outboxCh = make(chan any, inboxCapacity)
+		b.outbox[id.Uint32()] = outboxCh
 	}
-	b.pending[layerNum] = append(b.pending[layerNum], msg)
-	return nil
-}
 
-func (b *Broker) cleanOldLayers() {
-	b.mu.Lock()
-	defer b.mu.Unlock()
-
-	for i := b.minDeleted.Add(1); i.Before(b.latestLayer); i = i.Add(1) {
-		_, exist := b.outbox[i.Uint32()]
-
-		if !exist { // unregistered
-			b.minDeleted = b.minDeleted.Add(1)
-		} else { // encountered first still running layer
-			break
+	oversize := len(b.outbox) - b.limit
+	if oversize >= 0 {
+		// Remove earliest layers to make space for this layer
```

Comment:
i would leave a TODO here to clean up the consensus process that owns the earliest layer properly.

in essence, the oldest outbox removed will cause confusing error msgs for the ongoing consensus process. it is still running, but doesn't have a outbox to receive msgs.

Reply:
Right, I wondered about this. So we actually don't want any limit imposed by the broker?

Reply:
> So we actually don't want any limit imposed by the broker?

not sure what this mean.

current state is that the limit of concurrent consensus processes is ONLY imposed by the broker, and in a non-obvious way (by removing the inbox and hence dropping all communication).

i think we should do the following in described order)
- call `proc.terminate()` on some error
- remove its inbox

the logic should reside in `onTick()` in `hare.go` where it register new consensus process. if there are too many, remove and ask broker to remove inbox. the whole prune logic isn't even necessary in broker.

Reply:
Ok got you, in that case I think it would be clearer to only register with the broker if we are below some concurrent consensus process limit. And then we can free the broker from any knowledge of limits or pruning.